app/code/Magento/ConfigurableProduct/Model/ResourceModel/Product/Type/Configurable/Product/Collection.php

```diff
@@ -77,14 +77,14 @@ public function setProductFilter($product)
     protected function _renderFilters()
     {
         parent::_renderFilters();
-        $metadata = $this->getProductEntityMetadata();
-        $parentIds = [];
-        foreach ($this->products as $product) {
-            $parentIds[] = $product->getData($metadata->getLinkField());
+        if (!is_array($this->products)) {
+            return $this;
         }
-
+        $productEntityMetadata = $this->getProductEntityMetadata();
+        $parentIds = array_map(function (Product $product) use ($productEntityMetadata) {
```

Comment:
No reason to use `array_map` here.

Reply:
array_map results in less verbose code when you want to create a new array of same size. I see no reason not to use it tbh

Reply:
+ you gain type safety for array elements

Reply:
Code became much less simpler, please revert.

Type safety is not a real issue and does not justify decrease of readability.

Reply:
@orlangur  I disagree with that. It seems to be a very subjective opinion of you

1. array_map is much clearer in its intention: "map values of an array to a new array" .
2. type safety is always better in regards of potential future bugs
3. the type clearly states with which object your dealing (better readability)
4. with this awful value=[]; foreach... approach you have a code block that has to stay together, but is not bound together by syntax

Reply:
```
        $parentIds = [];	           
        foreach ($this->products as $product) {	
            $parentIds[] = $product->getData($metadata->getLinkField());	
        }
```

is MUCH easier to comprehend and there is nothing subjective with that. 

Please just fix a bug and revert anything unrelated. For using `array_map` instead of such a simple loop you can create a proposal in https://github.com/magento/architecture.

Reply:
it's subjective as long as you cannot explain, why it is easier

Reply:
+ it's not an architecture decision, it's just about clean, readable code

Reply:
for further reading: https://szymonkrajewski.pl/why-you-should-use-array-functions-instead-of-basic-loops/

Reply:
@davidverholen `foreach` I mentioned is a clean, readable code. Please add at least SOME argument why such `foreach` is bad. Your code is not clean and not readable at all.

There is nothing new for me in that article, I know all about such functions and there are good use cases for them but this is not the one.

If you want to perform such kind of useless refactoring, you need to get an architectural approval first and add recommendation somewhere in https://devdocs.magento.com/guides/v2.3/coding-standards/technical-guidelines.html which would require any suitable loop to be written in form of `array_map`.

This PR is a bugfix and your "refactoring" is not related to achieving a goal. That's why I'm asking you to revert unrelated changes and deliver suggested improvements in separate PR once they get approved.

Reply:
I provided several arguments why array_map is superior in that context, while you provided exactly 0